soundbay/inference.py

```diff
@@ -180,64 +153,62 @@ def infer_single_file(
-                                     pin_memory=False)
-
-        # predict
-        predict_prob = predict_proba(model, test_dataloader, device, None)
-
-        results_df = pandas.DataFrame(predict_prob)
-        if hasattr(test_dataset, 'metadata'):
-            concat_dataset = pandas.concat([test_dataset.metadata, results_df], axis=1) #TODO: make sure metadata column order matches the prediction df order
-            metrics_dict = Logger.get_metrics_dict(concat_dataset["label"].values.tolist(),
-                                                   np.argmax(predict_prob, axis=1).tolist(),
-                                                   results_df.values)
-            print(metrics_dict)
-        else:
-            concat_dataset = results_df
-            print("Notice: The dataset has no ground truth labels")
-
-        # create raven file
-        if save_raven:
-            all_channel_raven_list.append(
-                inference_csv_to_raven(results_df, predict_prob.shape[1], dataset_args['seq_length'], 1, threshold, 'call',
-                                       channel, dataset_args['data_sample_rate'] // 2)
-            )
-
-        # add to general inference result
-        concat_dataset.insert(0, 'channel', [channel + 1] * len(concat_dataset))
-        all_channel_list.append(concat_dataset)
+    test_dataset = datasets_dict[dataset_type](**dataset_args)
+    test_dataloader = DataLoader(dataset=test_dataset, shuffle=False, batch_size=batch_size, num_workers=0,
+                                 pin_memory=False)
+
+    # predict
+    predict_prob = predict_proba(model, test_dataloader, device, None)
+    label_names = ['Noise'] + [f'Call_{i}' for i in
+                               range(1, predict_prob.shape[1] + 1)] if label_names is None else label_names
+
+    results_df = pandas.DataFrame(predict_prob, columns=label_names)
+
+    concat_dataset = pandas.concat([test_dataset.metadata, results_df], axis=1)
+    # create raven file
+    if save_raven:
+        for file, df in concat_dataset.groupby('filename'):
+            file_raven_lists = []
+            for i in range(1, predict_prob.shape[1]):
+                file_raven_lists.append(
+                                       inference_csv_to_raven(results_df=df,
+                                                              num_classes=predict_prob.shape[1],
+                                                              seq_len=dataset_args['seq_length'],
+                                                              selected_class=label_names[i],
+                                                              threshold=threshold,
+                                                              class_name=label_names[i],
+                                                              max_freq=dataset_args['sample_rate'] // 2)
+                                   )
+            whole_file_df = pd.concat(file_raven_lists, axis=0).sort_values('Begin Time (s)')
```

Comment:
suggestion: first by file, and then by begin_time?

Reply:
same as next suggestion, resolving